### ISSUE TITLE
Difference between JUnit 4 and Jupiter when using rules and ExternalResourceSupport

### ISSUE THREAD

**Issue by BlueIce** [junit-team/junit5#1128]
## Overview

I noticed that in Jupiter tests a rule is not picked up when a method is annotated with `@Rule` and the return type is a subclass of `ExternalResource`. Here is an example test case for JUnit Jupiter:

```java
import org.junit.Rule;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.migrationsupport.rules.ExternalResourceSupport;
import org.junit.rules.ExternalResource;
import org.junit.rules.TestRule;

@ExtendWith(ExternalResourceSupport.class)
class RuleJUnit5Test {

    @Rule
    public ExampleExternalResourceRule getExampleExternalResourceRule() {
        return new ExampleExternalResourceRule("ExampleExternalResourceRule");
    }

    @Rule
    public ExternalResource getExternalResource() {
        return new ExampleExternalResourceRule("ExternalResource");
    }

    @Rule
    public TestRule getTestRule() {
        return new ExampleExternalResourceRule("TestRule");
    }

    @Test
    void test() {
    }

    public static class ExampleExternalResourceRule extends ExternalResource {

        private String identifier;

        ExampleExternalResourceRule(String identifier) {
            this.identifier = identifier;
            System.out.println(identifier + " init");
        }

        @Override
        protected void before() throws Throwable {
            System.out.println(identifier + " before");
        }

        @Override
        protected void after() {
            System.out.println(identifier + " after");
        }
    }

}
```

And an equivalent JUnit 4 test case :
```java
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExternalResource;
import org.junit.rules.TestRule;

public class RuleJUnit4Test {

    @Rule
    public ExampleExternalResourceRule getExampleExternalResourceRule() {
        return new ExampleExternalResourceRule("ExampleExternalResourceRule");
    }

    @Rule
    public ExternalResource getExternalResource() {
        return new ExampleExternalResourceRule("ExternalResource");
    }

    @Rule
    public TestRule getTestRule() {
        return new ExampleExternalResourceRule("TestRule");
    }

    @Test
    public void test() {
    }

    public static class ExampleExternalResourceRule extends ExternalResource {

        private String identifier;

        ExampleExternalResourceRule(String identifier) {
            this.identifier = identifier;
            System.out.println(identifier + " init");
        }

        @Override
        protected void before() throws Throwable {
            System.out.println(identifier + " before");
        }

        @Override
        protected void after() {
            System.out.println(identifier + " after");
        }
    }

}
```

Console output JUnit Jupiter:
```
ExternalResource init
ExternalResource before
TestRule init
TestRule before
TestRule init
TestRule after
ExternalResource init
ExternalResource after
```

Console output JUnit 4:
```
ExampleExternalResourceRule init
ExternalResource init
TestRule init
TestRule before
ExternalResource before
ExampleExternalResourceRule before
ExampleExternalResourceRule after
ExternalResource after
TestRule after
```

For one the rule with a return type of `ExampleExternalResourceRule` is not used in the Jupiter case. And secondly the rule methods are called twice.

**Comment by mmerdes** [comment]
@BlueIce 
Thanks for raising the issue. Will take a look - might take some time however.

**Comment by marcphilipp** [comment]
_in progress_

**Comment by marcphilipp** [comment]
Fixed by #1129.
